File: src/main/scala/com/vc/medium/L791.scala
package com.vc.medium

object L791 {
  def customSortString(S: String, T: String): String = {
    import scala.collection.mutable
    val map = new mutable.HashMap[Char, Int]()
    T.foreach(ch => map.put(ch, map.getOrElse(ch, 0) + 1))

    val str = new mutable.StringBuilder
    def appendTimes(ch:Char, times: Int): Unit = (0 until times).foreach(_ => str.append(ch))

    S.foreach(ch => {
      if(map.contains(ch)) {
        appendTimes(ch, map(ch))
        map -= ch
      }
    })
    map.foreach(e => appendTimes(e._1, e._2))
    str.toString
  }
}

File: src/main/java/com/vc/medium/L791.java
package com.vc.medium;

import java.util.*;

class L791 {
    public String customSortString(String S, String T) {
        HashMap<Character, Integer> map = new HashMap<>();
        for(int i = 0; i < T.length(); i++) {
            char ch = T.charAt(i);
            map.put(ch, map.getOrDefault(ch, 0) + 1);
        }
        StringBuilder str = new StringBuilder();
        for(int i = 0; i < S.length(); i++) {
            char ch = S.charAt(i);
            if(map.containsKey(ch)) {
                for(int j = 0; j < map.get(ch); j++) str.append(ch);
                map.remove(ch);
            }
        }
        for(Map.Entry<Character, Integer> entry: map.entrySet()) {
            for(int i = 0; i < entry.getValue(); i++) str.append(entry.getKey());
        }
        return str.toString();
    }
}
